### PR TITLE
🧮 Rearrange DPOTrainer

### PR DESCRIPTION
**Pull request by DaizeDong**
# What does this PR do?

Rearranged the code structure in `DPOTrainer`. Details include:
- **Separated tedious model preparation procedures into three functions:** `_create_model_from_path`, `_prepare_peft_model`, and `_prepare_gradient_checkpointing`, while keeping the original logic unchanged. This follows the implementation in [`SFTTrainer`](https://github.com/huggingface/trl/blob/main/trl/trainer/sft_trainer.py#L410), which also splits model preparation into various functional parts for clarity.
- **Grouped the model preparation code together.** This includes moving `disable_dropout_in_model(model)` and `model.warnings_issued["estimate_tokens"]` ahead so that all operations on the model are grouped together. This is safe and would not influence the execution.
- **Updated initialization for `model` and `processing_class`.** Now we can pass `None` as the value for `model` and `processing_class` in `DPOTrainer`. This feature is adapted from `SFTTrainer`.
- **Padding-free Checks.** Synced some checks from `SFTTrainer`.
- **Updated docs.** Now most docs align with `SFTTrainer`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
Thanks! I like it! Please let us know when it's ready for review!

**Comment by DaizeDong** [comment]
Thanks for your attention! It's ready for review. cc @qgallouedec

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -216,6 +216,7 @@ def __init__(
         preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
         peft_config: Optional[dict] = None,
     ):
+        # Model
         if model is None:
             raise ValueError("No model provided. Please provide a model to train.")
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:221]: You can remove this, and replace 

```diff
- model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
+ model:Union[PreTrainedModel, nn.Module, str],
```

**Comment by DaizeDong** [comment]
Thank you for your suggestions! I've updated the mentioned parts and more to sync the `DPOTrainer` with `SFTTrainer`!

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
Thanks! definitely helpful
